client/session.go

```diff
@@ -1420,27 +1437,45 @@ func (s *session) Truncate(namespace ident.ID) (int64, error) {
 	return truncated, resultErr.FinalError()
 }
 
-func (s *session) peersForShard(shard uint32) ([]peer, error) {
+type peers struct {
+	peers            []peer
+	selfExcluded     bool
+	selfHostShardSet topology.HostShardSet
+}
+
+func (s *session) peersForShard(shard uint32) (peers, error) {
+	var (
+		lookupErr error
+		result    = peers{peers: make([]peer, 0, s.topoMap.Replicas())}
+	)
+
 	s.RLock()
-	peers := make([]peer, 0, s.topoMap.Replicas())
 	err := s.topoMap.RouteShardForEach(shard, func(idx int, host topology.Host) {
 		if s.origin != nil && s.origin.ID() == host.ID() {
 			// Don't include the origin host
+			result.selfExcluded = true
+			// Include the origin host shard set for help determining quorum
+			hostShardSet, ok := s.topoMap.LookupHostShardSet(host.ID())
+			if !ok {
+				lookupErr = fmt.Errorf("could not find shard set for host ID: %s", host.ID())
+			}
+			result.selfHostShardSet = hostShardSet
 			return
 		}
-		peers = append(peers, newPeer(s, host))
+		result.peers = append(result.peers, newPeer(s, host))
 	})
 	s.RUnlock()
-	if err != nil {
-		return nil, err
+	if resultErr := xerrors.FirstError(err, lookupErr); resultErr != nil {
+		return peers{}, resultErr
 	}
-	return peers, nil
+	return result, nil
```

Comment:
Any concern about sharing this object outside the RLock? 

Main concern is if it can cause panics in weird situations: e.g. if we're bootstrapping during a topology change (for w/e reason).

Reply:
@prateek how would this cause a panic?

Reply:
Yeah this isn't an issue, that's why the `peer` interface only lets you borrow a connection (and if topology occurs and its no longer in the topology then borrowing connection will just fail):
```go
type peer interface {
	// Host gets the host
	Host() topology.Host

	// BorrowConnection will borrow a connection and execute a user function
	BorrowConnection(fn withConnectionFn) error
}
```